File: matchy/enrichment.py
from __future__ import annotations

from concurrent.futures import Future, ThreadPoolExecutor, TimeoutError, as_completed
import logging

from .models import EmailCandidate

LOGGER = logging.getLogger(__name__)


class EnrichmentMixin:
    #R015: Replace each candidate's body_text with the full email body fetched from Mailcart so that
    #R015: amount/keyword/compact-merchant hints can score against the real message body. Returns the
    #R015: original candidates unchanged when the feature flag is off or when the client is missing
    #R015: get_message (older Mailcart deployments). Per-candidate failures fall through to the
    #R015: original candidate so a flaky message id does not poison the whole run.
    def _enrich_candidate_bodies(self, candidates: list[EmailCandidate], transaction_id: str) -> list[EmailCandidate]:
        result = candidates
        config = self._body_enrichment_config(candidates)
        if config is not None:
            get_message, enrich_count, timeout_seconds, max_workers, per_message_timeout = config
            message_ids = self._unique_message_ids(candidates, enrich_count)
            payload_by_id = self._fetch_message_payloads(
                get_message, message_ids, transaction_id, max_workers, timeout_seconds, per_message_timeout,
            )
            result = self._apply_body_enrichment(candidates, enrich_count, payload_by_id)
        return result

    def _body_enrichment_config(self, candidates: list[EmailCandidate]):
        config = None
        enabled = bool(getattr(self._settings, "mailcart_body_enrichment_enabled", False))
        get_message = getattr(self._mailcart_client, "get_message", None)
        if candidates and enabled and callable(get_message):
            limit = int(getattr(self._settings, "mailcart_body_enrichment_limit", 75) or 75)
            timeout_seconds = max(1, int(getattr(self._settings, "mailcart_body_enrichment_timeout_seconds", 25) or 25))
            max_workers = max(1, int(getattr(self._settings, "mailcart_body_enrichment_max_workers", 8) or 8))
            per_message_timeout = max(1, int(getattr(self._settings, "mailcart_get_message_timeout_seconds", 6) or 6))
            enrich_count = min(len(candidates), limit)
            if enrich_count >= 1:
                config = (get_message, enrich_count, timeout_seconds, max_workers, per_message_timeout)
        return config

    def _unique_message_ids(self, candidates: list[EmailCandidate], enrich_count: int) -> list[str]:
        message_ids: list[str] = []
        for index in range(enrich_count):
            message_id = str(candidates[index].message_id)
            if message_id not in message_ids:
                message_ids.append(message_id)
        return message_ids

    def _fetch_message_payloads(self, get_message, message_ids, transaction_id,
                                max_workers, timeout_seconds, per_message_timeout) -> dict[str, dict]:
        payload_by_id: dict[str, dict] = {}
        future_to_message_id: dict[Future, str] = {}
        with ThreadPoolExecutor(max_workers=min(max_workers, len(message_ids))) as executor:
            for message_id in message_ids:
                future_to_message_id[executor.submit(get_message, message_id, per_message_timeout)] = message_id
            try:
                for future in as_completed(future_to_message_id, timeout=timeout_seconds):
                    message_id = future_to_message_id[future]
                    try:
                        payload_by_id[message_id] = future.result() or {}
                    except Exception as exc:
                        LOGGER.warning(
                            "mailcart get_message failed message_id=%s transaction_id=%s error=%s",
                            message_id, transaction_id, exc,
                        )
            except TimeoutError:
                unresolved_count = len([future for future in future_to_message_id if not future.done()])
                LOGGER.warning(
                    "mailcart body enrichment timed out transaction_id=%s unresolved_candidates=%s timeout_seconds=%s",
                    transaction_id, unresolved_count, timeout_seconds,
                )
        return payload_by_id

    def _enrichment_body_text(self, payload) -> str:
        body_text = ""
        if payload:
            body_text = (
                str(payload.get("text_body") or "").strip()
                or str(payload.get("html_body") or "").strip()
                or str(payload.get("body_text") or "").strip()
            )
        return body_text

    def _apply_body_enrichment(self, candidates, enrich_count, payload_by_id) -> list[EmailCandidate]:
        enriched: list[EmailCandidate] = []
        for index in range(enrich_count):
            candidate = candidates[index]
            payload = payload_by_id.get(str(candidate.message_id))
            body_text = self._enrichment_body_text(payload)
            if body_text:
                enriched.append(EmailCandidate(
                    message_id=candidate.message_id,
                    subject=candidate.subject or str(payload.get("subject") or ""),
                    preview=candidate.preview or str(payload.get("preview") or ""),
                    received_at=candidate.received_at,
                    sender=candidate.sender or str(payload.get("sender") or ""),
                    body_text=body_text,
                ))
            else:
                enriched.append(candidate)
        if len(candidates) > enrich_count:
            enriched.extend(candidates[enrich_count:])
        return enriched

    #R050: Scope matchable candidates to messages containing a standalone CLDR currency code or symbol.
    def _filter_currency_candidates(self, candidates: list[EmailCandidate]) -> list[EmailCandidate]:
        filtered = candidates
        matcher = getattr(self, "_cldr_currency_matcher", None)
        if candidates and matcher is not None and getattr(matcher, "tokens", frozenset()):
            filtered = []
            for candidate in candidates:
                text_blob = f"{candidate.subject} {candidate.preview} {candidate.body_text}"
                if matcher.contains_standalone_currency(text_blob):
                    filtered.append(candidate)
        return filtered

File: src/core/src/enrichment.cpp
#include "matchycore/enrichment.hpp"
#include <chrono>
#include <condition_variable>
#include <cstdio>
#include <map>
#include <mutex>
#include <thread>

namespace matchycore::enrichment
{ namespace
// #R001: Matchycore traceability implementation coverage.
 { std::string Strip(const std::string &value)
  { std::size_t begin = value.find_first_not_of(" \t\n\r\f\v");
   std::string out;
   if (begin != std::string::npos) out = value.substr(begin, value.find_last_not_of(" \t\n\r\f\v") - begin + 1);
   return out;
  }

// #R001: Matchycore traceability implementation coverage.
  std::string PayloadString(const nlohmann::json &payload, const std::string &key)
  { std::string value;
   if (payload.is_object() && payload.contains(key) && payload[key].is_string()) value = payload[key].get<std::string>();
   return value;
  }

  // Shared state for the bounded fetch pool; abandoned threads keep it (and the client) alive.
  class FetchState
  { public:
   std::mutex mutex;
   std::condition_variable cv;
   std::map<std::string, nlohmann::json> payload_by_id;
   std::vector<std::string> message_ids;
   std::size_t next_index = 0;
   std::size_t completed = 0;
   std::shared_ptr<mailcart::MailcartApi> client;
   int per_message_timeout = 0;
  };
 }

// #R001: Matchycore traceability implementation coverage.
 std::string EnrichmentBodyText(const nlohmann::json &payload)
 { std::string body_text = Strip(PayloadString(payload, "text_body"));
  if (body_text.empty()) body_text = Strip(PayloadString(payload, "html_body"));
  if (body_text.empty()) body_text = Strip(PayloadString(payload, "body_text"));
  return body_text;
 }

 std::vector<EmailCandidate> EnrichCandidateBodies(const std::shared_ptr<mailcart::MailcartApi> &client,
                                                   const Settings &settings,
                                                   const std::vector<EmailCandidate> &candidates,
                                                   const std::string &transaction_id)
 { std::vector<EmailCandidate> result = candidates;
  bool enabled = settings.mailcart_body_enrichment_enabled() && !candidates.empty() && client != nullptr;
  if (enabled)
  { int limit = settings.mailcart_body_enrichment_limit() != 0 ? settings.mailcart_body_enrichment_limit() : 75;
   int timeout_seconds = std::max(1, settings.mailcart_body_enrichment_timeout_seconds() != 0
    ? settings.mailcart_body_enrichment_timeout_seconds() : 25);
   int max_workers = std::max(1, settings.mailcart_body_enrichment_max_workers() != 0
    ? settings.mailcart_body_enrichment_max_workers() : 8);
   int per_message_timeout = std::max(1, settings.mailcart_get_message_timeout_seconds() != 0
    ? settings.mailcart_get_message_timeout_seconds() : 6);
   std::size_t enrich_count = std::min(candidates.size(), static_cast<std::size_t>(std::max(1, limit)));
   // #R001: Preserve first-seen ordering while deduplicating ids within the enrichment window.
   auto state = std::make_shared<FetchState>();
   state->client = client;
   state->per_message_timeout = per_message_timeout;
   for (std::size_t index = 0; index < enrich_count; index += 1)
   { const std::string &message_id = candidates[index].message_id();
    bool seen = false;
    for (const std::string &existing : state->message_ids)
     if (existing == message_id) seen = true;
    if (!seen) state->message_ids.push_back(message_id);
   }
   // #R001: Fetch payloads concurrently, tolerating per-message failures and the overall deadline.
   std::size_t worker_count = std::min<std::size_t>(static_cast<std::size_t>(max_workers), state->message_ids.size());
   for (std::size_t w = 0; w < worker_count; w += 1)
   { std::thread([state, transaction_id]()
    { bool working = true;
     while (working)
     { std::string message_id;
      { std::lock_guard<std::mutex> lock(state->mutex);
       if (state->next_index < state->message_ids.size())
       { message_id = state->message_ids[state->next_index];
        state->next_index += 1;
       }
       else working = false;
      }
      if (working)
      { nlohmann::json payload = nlohmann::json::object();
       try
       { payload = state->client->GetMessage(message_id, state->per_message_timeout);
       }
       catch (const std::exception &exc)
       { std::fprintf(stderr, "mailcart get_message failed message_id=%s transaction_id=%s error=%s\n",
                      message_id.c_str(), transaction_id.c_str(), exc.what());
       }
       std::lock_guard<std::mutex> lock(state->mutex);
       state->payload_by_id[message_id] = payload;
       state->completed += 1;
       state->cv.notify_all();
      }
     }
    }).detach();
   }
   std::map<std::string, nlohmann::json> payload_by_id;
   { std::unique_lock<std::mutex> lock(state->mutex);
    bool finished = state->cv.wait_for(lock, std::chrono::seconds(timeout_seconds),
                                       [&state]() { return state->completed >= state->message_ids.size(); });
    if (!finished)
     std::fprintf(stderr, "mailcart body enrichment timed out transaction_id=%s unresolved_candidates=%zu timeout_seconds=%d\n",
                  transaction_id.c_str(), state->message_ids.size() - state->completed, timeout_seconds);
    payload_by_id = state->payload_by_id;
   }
   // #R001: Apply enriched payloads to the configured candidate prefix, leaving unmatched rows unchanged.
   std::vector<EmailCandidate> enriched;
   for (std::size_t index = 0; index < enrich_count; index += 1)
   { const EmailCandidate &candidate = candidates[index];
    nlohmann::json payload = nlohmann::json::object();
    auto found = payload_by_id.find(candidate.message_id());
    if (found != payload_by_id.end()) payload = found->second;
    std::string body_text = EnrichmentBodyText(payload);
    if (!body_text.empty())
     enriched.emplace_back(candidate.message_id(),
                           candidate.subject().empty() ? PayloadString(payload, "subject") : candidate.subject(),
                           candidate.preview().empty() ? PayloadString(payload, "preview") : candidate.preview(),
                           candidate.received_at(),
                           candidate.sender().empty() ? PayloadString(payload, "sender") : candidate.sender(),
                           body_text);
    else enriched.push_back(candidate);
   }
   for (std::size_t index = enrich_count; index < candidates.size(); index += 1) enriched.push_back(candidates[index]);
   result = enriched;
  }
  return result;
 }

// #R001: Matchycore traceability implementation coverage.
 std::vector<EmailCandidate> FilterCurrencyCandidates(const cldr::CldrCurrencyMatcher &matcher,
                                                      const std::vector<EmailCandidate> &candidates)
 { std::vector<EmailCandidate> filtered = candidates;
  if (!candidates.empty() && !matcher.tokens().empty())
  { filtered.clear();
   for (const EmailCandidate &candidate : candidates)
   { std::string text_blob = candidate.subject() + " " + candidate.preview() + " " + candidate.body_text();
    if (matcher.ContainsStandaloneCurrency(text_blob)) filtered.push_back(candidate);
   }
  }
  return filtered;
 }
}
